protocol/communities/community_event.go

```diff
@@ -0,0 +1,430 @@
+		AcceptedRequestsToJoin: changes.AcceptedRequestsToJoin,
+	}
+}
+
+func (o *Community) ToCommunityRequestToJoinRejectCommunityEvent(changes *CommunityEventChanges) *CommunityEvent {
+	return &CommunityEvent{
+		CommunityEventClock:    o.NewCommunityEventClock(),
+		Applied:                true,
+		Type:                   protobuf.CommunityEvent_COMMUNITY_REQUEST_TO_JOIN_REJECT,
+		RejectedRequestsToJoin: changes.RejectedRequestsToJoin,
+	}
+}
+
+func (o *Community) UpdateCommuntyByEvents(communityEventMessage *CommunityEventsMessage) (*CommunityChanges, error) {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	if communityEventMessage.CommunityDescriptionClock != o.config.CommunityDescription.Clock {
+		return nil, errors.New("clock for admin event message is outdated")
+	}
+
+	// Merge community admin events to existing community. Admin events must be stored to the db
+	// during saving the community
+	o.mergeCommunityEvents(communityEventMessage.Events)
+
+	// Create a deep copy of current community so we can update CommunityDescription by new admin events
+	copy := o.createDeepCopy()
+
+	// Update the copy of the Community with the new community events
+	err := copy.updateCommuntyDescriptionByNewEvents()
+	if err != nil {
+		return nil, err
+	}
+
+	// Collect `CommunityChanges` data by searching a difference between `CommunityDescrption`
+	// with applied events and `CommunityDescrption` with new events
+	changes, err := o.collectCommuntyChanges(copy.config.CommunityDescription)
+	if err != nil {
+		return nil, err
+	}
+
+	changes.Community = copy
+
+	return changes, nil
+}
+
+func (o *Community) updateCommuntyDescriptionByNewEvents() error {
+	prevAdminClock := o.config.CommunityDescription.Clock + 1
+	for i := range o.config.Events {
+		communityEvent := &o.config.Events[i]
+		if !communityEvent.Applied {
+			err := validateCommunityEvent(communityEvent)
+			if err != nil {
+				return err
+			}
+			err = o.updateCommuntyDescriptionByCommunityEvent(*communityEvent, prevAdminClock)
+			if err != nil {
+				return err
+			}
+		}
+		prevAdminClock = communityEvent.CommunityEventClock
+		communityEvent.Applied = true
+	}
+	return nil
+}
+
+func (o *Community) updateCommuntyDescriptionByAppliedEvents() error {
+	prevAdminClock := o.config.CommunityDescription.Clock + 1
+	for _, communityEvent := range o.config.Events {
+		if communityEvent.Applied {
+			err := o.updateCommuntyDescriptionByCommunityEvent(communityEvent, prevAdminClock)
+			if err != nil {
+				return err
+			}
+		}
+		prevAdminClock = communityEvent.CommunityEventClock
+	}
+	return nil
+}
+
+func (o *Community) updateCommuntyDescriptionByCommunityEvent(communityEvent CommunityEvent, prevAdminClock uint64) error {
+	if communityEvent.CommunityEventClock <= o.config.CommunityDescription.Clock {
+		return errors.New("clock for admin event is outdated")
+	}
+
+	if !(prevAdminClock == communityEvent.CommunityEventClock || prevAdminClock+1 == communityEvent.CommunityEventClock) {
```

Comment:
it seems like we expect events to be strictly in sequence, but unfortunately that's not possible, since there's no consensus mechanism,which might lead to inconsistencies.

For example:

1) admin-1 publishes event admin-1-event with clock 2
2) admin-2 at the same time publish admin-2-event with clock 2

Client-1 will receive 1 first, and process it, discarding 2.
Client-2 will receive 2 first, and process it, discarding 1.

Did I get that right that this is what the code would do?

In this case, the correct behavior I believe is for Client-1 to process both events (in a pre-established order, in this case you want to break ties deterministically, say by ordering them by clock-pk), as they are not causally related. We need to think of potential conflict, those can be sorted out by some arbitrary rules (i.e if admin-1 makes X an admin and admin-2 removes X from the community, all the clients will agree on what's the outcome based on some rules we decide).

Reply:
We merge all admin events in one list, so if admin-1 and admin-2 will publish 2 different events with clock 2, clients will merge them and update ([logic inside](https://github.com/status-im/status-go/pull/3678/files#diff-837cc1410fc2cdd0d31b2a9c8b2e03f37ae09a824a55261fb70b2aaffeaf4513R100-R115))
In your case, Client-1 and Client-2 will megre both admin events and apply both

The idea of this check was not to allow apply events in order: 1-2-4, where 3 is unknown, but now I looked at this case from another corner and I think it is impossible as all new admin events will be +1 from previous or from CommunityDescription clock